content/blog/2023/2023-01-17-trubar.md

```diff
@@ -0,0 +1,95 @@
+
+The above example uses the old-style interpolation with `format`. is does not work for Python's f-strings. With f-strings, we would have `print(tr(f"Data contains {n} instances")`. The number, `n`, is interpolated before the function call, so `tr` already receives the entire string, like `"Data contains 1234 instances"`. This obviously doesn't match any pre-translated messages, because it includes a specific number. Call `tr` earlier? See, that's the thing: you can't. There is no such string as `"Data contains {n} instances"`. It is never "materialized": in Python's abstract syntax tree, these are three distinct elements (a constant `"Data contains "`, expression `n`, and a constant ` instance`). At the moment they are joined into an actual object (of type `str`) that can be passed to some function, the value of `n` is already interpolated.
+
+Gettext-like approaches thus can't work on f-strings. To our knowledge, there is currently no tool to support translation of f-strings.
+
+Luckily, we are programmers.
+
+## Enters Trubar
+
+We developed Trubar. Named after a colorful translator and publisher from Slovenian history, Trubar collects all strings from sources and places them into a yaml-like file. There, they can be translated - or marked as strings that must not be translated. Then, Trubar copies the sources, substituting strings with their translations, if provided.
+
+This is a part of the file related to the translation of the Table widget above.
+
+```yaml
+widgets/data/owtable.py:
+    class `OWDataTable`:
+        Data Table: Tabela
+        icons/Table.svg: false
+        class `Inputs`:
+            Data: Podatki
+        class `Outputs`:
+            Selected Data: Izbor podatkov
+        def `__init__`:
+            Variables: Spremenljivke
+            show_attribute_labels: false
+            Show variable labels (if present): Pokaži oznake spremenljivk
+            show_distributions: false
+            Visualize numeric values: Vizualiziraj številske vrednosti
+            color_by_class: false
+            Color by instance classes: Obarvaj primere glede na razred
+            Selection: Izbor
+            select_rows: false
+            Select full rows: Izbiraj cele vrstice
+            Restore Original Order: Izvirni vrstni red
+        def `set_dataset`:
+            name: false
+            Data: Podatki
+```
+
+If it looks simple, it's because [it *is* simple](http://janezd.github.io/trubar/getting-started/).
+
+Besides working with f-strings, the advantage of Trubar's approach is that is does not pollute the sources with calls and parentheses. There's enough of them already. Instead of marking strings for translation within the source code, they are marked (for non-translation) in the message file. This must be done at some point, but it's better to do it in a dedicated place, not in the code.
```

Comment:
"The advantage of Trubar's approach is that is does not pollute the sources with calls and parentheses."
As gettext does?

Reply:
Yes. I now added "- unlike gettext -". This sentence used to be closer to description of gettext, but now too much other content came in between so a reminder is needed.